### ISSUE TITLE
Category\Collection::joinUrlRewrite should use the store set on the collection

### ISSUE THREAD

**Issue by alepane21** [magento/magento2#13704]
### Preconditions

1. Magento 2.1.8
2. PHP 7
3. MySQL 5.6

### Steps to reproduce

    $categoryCollection = $this->categoryCollectionFactory->create();
    $categoryCollection->setStoreId(1);
    $categoryCollection->addUrlRewriteToResult();
    echo $categoryCollection->getUrl();

### Expected result

1. should return the url of category in store 1

### Actual result

1. Instead it returns the url of default store

The problem is that Magento\Catalog\Model\ResourceModel\Category\Collection::joinUrlRewrite always use the store id from the store manager. I think that it should instead use the storeId set on the actual collection.

**Comment by orlangur** [comment]
Hi @alepane21, from method implementation

```
public function getStoreId()
    {
        if ($this->_storeId === null) {
            $this->setStoreId($this->_storeManager->getStore()->getId());
        }
        return $this->_storeId;
    }
```

it looks like proposed change could be applied safely.

Could you please prepare a pull request containing suggested change and appropriate test?

**Comment by alepane21** [comment]
I'm working on it.

**Comment by alepane21** [comment]
If the pull request pass on develop-2.2 I'm also going to do the backport to 2.1-develop.

**Comment by magento-engcom-team** [comment]
Hi @alepane21. Thank you for your report. 
The issue has been fixed in magento/magento2#13716 by @alepane21 in 2.2-develop branch
Related commit(s):
 - [fc73ae2b628392f4f4538e701f9b06f38ec5c9f8](https://github.com/magento/magento2/commit/fc73ae2b628392f4f4538e701f9b06f38ec5c9f8)
 - [8ec6e059a0160c0a2392fc820233d4e3cd93aa72](https://github.com/magento/magento2/commit/8ec6e059a0160c0a2392fc820233d4e3cd93aa72)
 - [0241a1c284cf0530d7dc62f9bbe531ba840867f1](https://github.com/magento/magento2/commit/0241a1c284cf0530d7dc62f9bbe531ba840867f1)
 - [3320fecbeffd9e793a205a4bfd29767cc38bc999](https://github.com/magento/magento2/commit/3320fecbeffd9e793a205a4bfd29767cc38bc999)
 - [c3ea127c6b257d72967ffc834b549410619ccc92](https://github.com/magento/magento2/commit/c3ea127c6b257d72967ffc834b549410619ccc92)
 - [eb04ab074cc52f5c2f5cc486fab014fca6b042da](https://github.com/magento/magento2/commit/eb04ab074cc52f5c2f5cc486fab014fca6b042da)

The fix will be available with the upcoming 2.2.5 release.

**Comment by magento-engcom-team** [comment]
Hi @alepane21. Thank you for your report. 
The issue has been fixed in magento/magento2#14381 by @rostyslav-hymon in 2.3-develop branch
Related commit(s):
 - [29431e958855ef19cc5289f8c86f39e2bc30f0a5](https://github.com/magento/magento2/commit/29431e958855ef19cc5289f8c86f39e2bc30f0a5)
 - [47c8d54a84770286e9912b83cf11805d29bf40ea](https://github.com/magento/magento2/commit/47c8d54a84770286e9912b83cf11805d29bf40ea)
 - [2a014892b5d19c04f5462d2f00c3c7f5b557b0c2](https://github.com/magento/magento2/commit/2a014892b5d19c04f5462d2f00c3c7f5b557b0c2)
 - [2d12dac1672aa4125b1012e454a8b58bb9ee9b73](https://github.com/magento/magento2/commit/2d12dac1672aa4125b1012e454a8b58bb9ee9b73)
 - [8a097fc1a83acd63b702661e12bcb10a00b36630](https://github.com/magento/magento2/commit/8a097fc1a83acd63b702661e12bcb10a00b36630)
 - [af3804ac4e846d42593ab73df3535a24d4ff9f8b](https://github.com/magento/magento2/commit/af3804ac4e846d42593ab73df3535a24d4ff9f8b)
 - [8220e3ff214d8911ba7dde971ed6856272228ca1](https://github.com/magento/magento2/commit/8220e3ff214d8911ba7dde971ed6856272228ca1)

The fix will be available with the upcoming 2.3.0 release.
